### ISSUE TITLE
API deprecate date_parser, add date_format

### ISSUE THREAD

**Issue by MarcoGorelli** [pandas-dev/pandas#50601]
**TLDR** the conversation here goes on for a bit, but to summarise, the suggestion is:

- deprecate `date_parser`, because it always hurts performance (counter examples welcome!)
- add `date_format`, because that can boost performance
- for anything else, amend docs to make clear that users should first in the data as `object`, and then apply their parsing

Performance-wise, this would only be an improvement to the status quo

------

As far as I can tell, `date_parser` is a net negative and only ever slows things down

In the best case, it only results in a slight degradation:
```python
timestamp_format = '%Y-%d-%m %H:%M:%S'

date_index = pd.date_range(start='1900', end='2000')

dates_df = date_index.strftime(timestamp_format).to_frame(name='ts_col')
data = dates_df.to_csv()
```

```python
In [6]: %%timeit
   ...: df = pd.read_csv(io.StringIO(data),
   ...:     date_parser=lambda x: pd.to_datetime(x, format=timestamp_format),
   ...:     parse_dates=['ts_col']
   ...: )
   ...: 
   ...: 
111 ms ± 3.02 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

In [7]: %%timeit
   ...: df = pd.read_csv(io.StringIO(data),
   ...:     #date_parser=lambda x: pd.to_datetime(x, format=timestamp_format),
   ...:     #parse_dates=['ts_col']
   ...: )
   ...: df['ts_col'] = pd.to_datetime(df['ts_col'], format=timestamp_format)
   ...: 
   ...: 
75.8 ms ± 1.98 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

Parsing element-by-element is also slower than just using `.apply`:
```python
In [21]: %%timeit
    ...: df = pd.read_csv(io.StringIO(data),
    ...:     #date_parser=lambda x: pd.to_datetime(x, format=timestamp_format),
    ...:     #parse_dates=['ts_col']
    ...: )
    ...: df['ts_col'].apply(lambda x: du_parse(x, dayfirst=True))
    ...: 
    ...: 
1.13 s ± 33.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

In [22]: %%timeit
    ...: df = pd.read_csv(io.StringIO(data),
    ...:     date_parser=lambda x: du_parse(x, dayfirst=True),
    ...:     parse_dates=['ts_col']
    ...: )
    ...: 
    ...: 
1.19 s ± 3.43 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

In the worst case, it results in **65x performance degradation**, see https://github.com/pandas-dev/pandas/pull/50586#issuecomment-1373473579 (and this gets way worse for larger datasets)

My suggestion is:
- deprecate `date_parser`
- introduce `date_format`, which _would_ actually deliver a performance improvement:

```python
In [1]: timestamp_format = '%Y-%d-%m %H:%M:%S'
   ...: 
   ...: date_index = pd.date_range(start='1900', end='2000')
   ...: 
   ...: dates_df = date_index.strftime(timestamp_format).to_frame(name='ts_col')
   ...: data = dates_df.to_csv()

In [2]: %%timeit
   ...: df = pd.read_csv(io.StringIO(data),
   ...:     date_format=timestamp_format,
   ...:     parse_dates=['ts_col']
   ...: )
   ...: 
   ...: 
19.5 ms ± 397 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

That's over 3 times as fast!

**Comment by sm-Fifteen** [comment]
The main drawback I can see of switching from a callback to a format string is that the callback enables you to use `Series.name` to use different patterns for each date column, though that can be remedied by allowing a dict to map column names to date formats.

Is there concern that the other parameters of to_datetime may end up having to be added to `read_csv`, like `utc=`, `unit=`, or `origin=`, or are these not a problem for datetime column parsing (at read time, anyway)?

At any rate, I certainly feel that this would be a pretty sizable improvement to date parsing, not just in terms of performance gains but also in terms of usability.

**Comment by MarcoGorelli** [comment]
Thanks for your input, much appreciated

how about `to_datetime_kwargs`? then `dayfirst` could be deprecated as well. so then someone can run
```
read_csv(data, parser_dates=['ts_col'], to_datetime_kwargs={'format': datetime_format, 'utc': False, 'dayfirst': True})
```
and it would be performant, because `to_datetime` wouldn't need to be called twice in

https://github.com/pandas-dev/pandas/blob/3bc2203b05ffbf3d18e912ce818f94bbf7820474/pandas/io/parsers/base_parser.py#L1133-L1135

, the kwargs could directly go to

https://github.com/pandas-dev/pandas/blob/3bc2203b05ffbf3d18e912ce818f94bbf7820474/pandas/io/parsers/base_parser.py#L1124-L1130

**Comment by MarcoGorelli** [comment]
> The main drawback I can see of switching from a callback to a format string is that the callback enables you to use Series.name to use different patterns for each date column

not sure I see what you mean - is this currently supported? do you have an example please?

**Comment by sm-Fifteen** [comment]
> > The main drawback I can see of switching from a callback to a format string is that the callback enables you to use Series.name to use different patterns for each date column
> 
> not sure I see what you mean - is this currently supported? do you have an example please?

Something like this:

```py
import pandas as pd
from io import StringIO
from tempfile import TemporaryFile

date_euro = pd.date_range(start='1900', end='2000', freq='1D').strftime("%d/%m/%Y %H:%M:%S").to_list()
date_us = pd.date_range(start='1900', end='2000', freq='1D').strftime("%m/%d/%Y %H:%M:%S").to_list()

date_df_pre = pd.DataFrame({
    'date_euro': date_euro,
    'date_us': date_us,
})

csv_buf = StringIO()
date_df_pre.to_csv(csv_buf)
csv_buf.seek(0)

date_df_pre
```

(day first vs month-first)

```                 date_euro              date_us
0      01/01/1900 00:00:00  01/01/1900 00:00:00
1      02/01/1900 00:00:00  01/02/1900 00:00:00
2      03/01/1900 00:00:00  01/03/1900 00:00:00
3      04/01/1900 00:00:00  01/04/1900 00:00:00
4      05/01/1900 00:00:00  01/05/1900 00:00:00
...                    ...                  ...
36520  28/12/1999 00:00:00  12/28/1999 00:00:00
36521  29/12/1999 00:00:00  12/29/1999 00:00:00
36522  30/12/1999 00:00:00  12/30/1999 00:00:00
36523  31/12/1999 00:00:00  12/31/1999 00:00:00
36524  01/01/2000 00:00:00  01/01/2000 00:00:00

[36525 rows x 2 columns]
```

```py
from typing import Union

def parse_all_dates(col: Union[str,pd.Series]):
    if not isinstance(col, pd.Series): return col
    if not pd.api.types.is_string_dtype(col.dtype): return col

    if col.name == 'date_euro': return pd.to_datetime(col, format="%d/%m/%Y %H:%M:%S")
    if col.name == 'date_us': return pd.to_datetime(col, format="%m/%d/%Y %H:%M:%S")
    return col

new_df = pd.read_csv(csv_buf, usecols=['date_euro', 'date_us'], parse_dates=['date_euro', 'date_us'], date_parser=parse_all_dates)
new_df
```

```
       date_euro    date_us
0     1900-01-01 1900-01-01
1     1900-02-01 1900-01-02
2     1900-03-01 1900-01-03
3     1900-04-01 1900-01-04
4     1900-05-01 1900-01-05
...          ...        ...
36520 1999-12-28 1999-12-28
36521 1999-12-29 1999-12-29
36522 1999-12-30 1999-12-30
36523 1999-12-31 1999-12-31
36524 2000-01-01 2000-01-01

[36525 rows x 2 columns]
```

A bit contrived, but I'm sure there are real-world use-cases that would need something like this, where different date formats are used per column.

**Comment by MarcoGorelli** [comment]
IIUC that's not currently supported by `date_parser` either right?

if so, let's keep that to a separate discussion (you're welcome to open a new issue and I'll gladly take a look)

**Comment by sm-Fifteen** [comment]
> Thanks for your input, much appreciated
> 
> how about `to_datetime_kwargs`? then `dayfirst` could be deprecated as well. so then someone can run

I can't say, I haven't had to use these parameters before and can't tell whether they're actually useful for CSV parsing, I'm just pointing it out for the record, in case there's a use case with these that I'm missing.

If the other args for to_datetime actually matter for file parsing, then I guess kwargs is probably the most extensible and least "future-intrusive" solution.

**Comment by sm-Fifteen** [comment]
> IIUC that's not currently supported by `date_parser` either right?
> 
> if so, let's keep that to a separate discussion (you're welcome to open a new issue and I'll gladly take a look)

Oh, no, that code snippet actually works and uses `date_parser` to parse both column differently. It strikes me as a potentially useful feature (on paper anyway, I can,t recall ever needing to do this) that would be solt is date_format were to only accept a flat string, and not, say, a dict of formats mapped by column names as well.

**Comment by MarcoGorelli** [comment]
ah I see, sorry, I should've actually tried running it. and if it's possible, then there's definitely someone out there using it 😄 

a dict mapping column names to kwargs would allow for that functionality to be kept though, right? like
```python
to_datetime(data, to_datetime_kwargs={'date_euro': {'dayfirst': True}, 'date_us': {'dayfirst': False}}
```
So, `to_datetime_kwargs` could either be a dict, or a dict of dicts

**Comment by sm-Fifteen** [comment]
I can't speak for the ergonomics of such a solution (the list of parameters for read_csv is already pretty long and complex), but it would at least preserve the functionnality of the callback, as far as I can tell. Looking at the current docs, there are cases where specific use of to_datetime is needed (such as [mixed timezone columns](https://pandas.pydata.org/docs/user_guide/io.html#io-csv-mixed-timezones)), so that wouldn't be lost, at least.

***

Truth be told, the more I look at the extended use-cases, the less I'm confident about having this sort of highly tunable functionnality (date parsing kwargs, that is) be baked directly in the read_csv function instead of having the user do the conversion as a second pass. [Pola-rs, for instance](https://pola-rs.github.io/polars-book/user-guide/howcani/timeseries/parsing_dates_times.html), only has a switch for automatic date parsing where it [tries to figure out your date format in an undocumented way based on the first value of a column](https://github.com/pola-rs/polars/blob/5f3e332fb2a653064f083b02949c527e0ec0afda/polars/polars-time/src/chunkedarray/utf8/mod.rs#L115-L146) by testing [a bunch of big-endian date formats](https://github.com/pola-rs/polars/blob/5f3e332fb2a653064f083b02949c527e0ec0afda/polars/polars-time/src/chunkedarray/utf8/mod.rs#L25-L63), and the guide saying you should leave that switch off and do the string-to-date conversion yourself if you have anything more complicated than that.

Leaving `read_csv` with just the simple knobs (i.e: just `parse_dates=` + `date_format=` and that's it) and having the doc reccomend a two-pass approach for the rest (like I was doing in #35296) might be preferable, but I'm far from having sufficent knowledge of the user expectations for Pandas to tell if that would be a good change or not.

**Comment by MarcoGorelli** [comment]
Thanks for your thoughts

Yes, I think you're right, I'd be +1 for simple knobs, and doc change

So, to summarise:

- deprecate `date_parser`, because it always hurts performance (counter examples welcome!)
- add `date_format`, because that can boost performance
- for anything else, amend docs to make clear that users should first in the data as `object`, and then apply their parsing

Performance-wise, this would only be an improvement to the status quo

**Comment by lithomas1** [comment]
+1 to this.

**Comment by jbrockmendel** [comment]
> to_datetime_kwargs

for use cases like the different-behavior-per-column id rather tell users to handle that after read_csv.  read_csv is way too complicated already.

**Comment by MarcoGorelli** [comment]
Agreed

I just checked, and if the format can be guessed, then just `parse_dates` is enough to offer a speed-up:
```python
In [2]: %%timeit
   ...: df = pd.read_csv(io.StringIO(data), parse_dates=['ts_col'])
   ...: 
   ...: 
18.2 ms ± 574 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

So, I'd suggest going further - just deprecate `date_parser`, and don't add anything new. For anything non-standard, users can read in the file and then apply `to_datetime`

Since PDEP4, the original issue which motivated adding the `date_parser` argument isn't relevant anymore anyway

**Comment by sm-Fifteen** [comment]
Leaving out non-automatic date parsing is something I'd normally be in favor of, but I should point out that manual date conversion [gets pretty convoluted](https://stackoverflow.com/a/63345157/12189790) when it comes to index columns and expecially MultiIndex.

```py
# Roughly copied from the SO answer above
# For a MultiIndex
df = pd.read_csv(infile, parse_dates={'mydatetime': ['date', 'time']}, index_col=['mydatetime', 'num'])

idx_mydatetime = df.index.get_level_values('mydatetime')
idx_num = df.index.get_level_values('num')
idx_mydatetime = pd.to_datetime(idx_mydatetime, exact=True, cache=True, format='%Y-%m-%d %H:%M:%S')
df.index = pd.MultiIndex.from_arrays([idx_mydatetime, idx_num])
```

There might be simpler ways of handling this in a "post-parsing" manner that I'm not aware of, but that,s still roughly what I'd imagine index parsing would look like if `read_csv` ends up with no date parsing utilities besides `parse_dates=`. Pola-rs can get away with this because they don't have indexes on their dataframes (the underlying Arrow dataframes don't either), but in Pandas, index manipulation unfortunately comes with extra steps like these, which the documentation would need to account for and might cause some amount of friction for users.

**Comment by MarcoGorelli** [comment]
I think that code looks fine, and rare-enough that to not need a special place in the documentation

**Comment by sm-Fifteen** [comment]
> rare-enough that to not need a special place in the documentation

Time-series data with an unconventionnal/ambiguous date format and one or more extra key columns (sensor name, histogram bucket, dataset name if there's multiple reports in one file, etc.)? I personally see these quite a lot; ISO 8601 is unfortunately not as universal as I would like. This is *CSV* we're talking about, people can't even agree whether the "C" stands for "Tab", "Semi-Colon" or "Space". Date formats are anybody's guess.

**Comment by MarcoGorelli** [comment]
Yes, and if it's not a format which pandas can guess, then people can read in the column and then run `to_datetime` with their given format

Do you have an example of something that people can currently do with `date_parser`, which would become more difficult without `date_parser`? Because the example you posted above 
```python
df = pd.read_csv(infile, parse_dates={'mydatetime': ['date', 'time']}, index_col=['mydatetime', 'num'])
```
doesn't use `date_parser`, and so would continue working

`parse_dates` is here to stay, I'm only suggesting that `date_parser` be deprecated

**Comment by sm-Fifteen** [comment]
> Because the example you posted above
> 
> ```python
> df = pd.read_csv(infile, parse_dates={'mydatetime': ['date', 'time']}, index_col=['mydatetime', 'num'])
> ```
> 
> doesn't use `date_parser`, and so would continue working

On its own, yes, but I would imagine most people have it written like this instead:

```py
df = pd.read_csv(
    infile, parse_dates={'mydatetime': ['date', 'time']}, index_col=['mydatetime', 'num'],
    date_parser=lambda x: pd.to_datetime(x, format=timestamp_format)
)
```

...and converting it to work without `date_parser` or a `date_format` replacement would mean adding the extra 4 lines I quoted above, which I'm sure most users would see as a downgrade in that regard.

```py
df = pd.read_csv(infile, parse_dates={'mydatetime': ['date', 'time']}, index_col=['mydatetime', 'num'])

idx_mydatetime = df.index.get_level_values('mydatetime')
idx_num = df.index.get_level_values('num')
idx_mydatetime = pd.to_datetime(idx_mydatetime, format=timestamp_format)
df.index = pd.MultiIndex.from_arrays([idx_mydatetime, idx_num])
```

Meanwhile, if `date_format` is introduced as a replacement for simple cases:

```py
df = pd.read_csv(
    infile,
    parse_dates={'mydatetime': ['date', 'time']},
    date_format={'mydatetime': timestamp_format},
    index_col=['mydatetime', 'num']
)
```

Which would cover the vast majority of uses of `date_parser` and make the transition a lot smoother for users.

**Comment by staticdev** [comment]
@MarcoGorelli I am struggling a bit to convert Pandas 2.0 recommended way.

I have the following (full code [here](https://github.com/staticdev/irpf-investidor/blob/main/src/irpf_investidor/report_reader.py#L106)):

```python
import datetime
import pandas as pd
def date_parse(value: str) -> datetime.datetime:
    return datetime.datetime.strptime(value.strip(), "%d/%m/%y")

df = pd.read_excel(
      filename,
      usecols="B:K",
      parse_dates=["Data Negócio"],
      date_parser=date_parse,
      skipfooter=4,
      skiprows=10,
  )
```

In this case, my final dataframe has a column "Data Negócio" with type `datetime64[ns]`. But if change it to:

```python
import datetime
import pandas as pd
DATE_FORMAT = "%d/%m/%y"

df = pd.read_excel(
      filename,
      usecols="B:K",
      parse_dates=["Data Negócio"],
      date_format=DATE_FORMAT,
      skipfooter=4,
      skiprows=10,
  )
```

My column now is an object and nothing is formatted. I get values such as '   01/12/21'. Is it a bug, or what am I missing?

**Comment by MarcoGorelli** [comment]
hi @staticdev - could you share a reproducible example please?

**Comment by MarcoGorelli** [comment]
(and also open a new issue if this is a bug - thanks!)

**Comment by ciscorucinski** [comment]
There doesn't seem to be a lot of documentation for users that have a working `date_parser` implementation and are running into the `FutureWarning` and need to go the `object` route; especially if they don't use pandas much.  

The issue I am having is that `date_format` is insufficient in my case because I unfortunately have multiple date formats in the dataset and it will only parse one or the other.  
1) `YYYY-MM-DD HH:MM:SS+HH:MM` - 2023-05-25 17:29:44+02:00
2) `YYYY-MM-DDTHH:MM:SS+HH:MM` - 2023-05-26T11:07:14+02:00


I am using a custom parser that uses `pd.Timestamp` which also nicely handled timezone conversion to UTC +00:00.

```
df = pd.read_csv(data, sep=",", parse_dates=['designation_date'], date_parser=lambda dt: pd.Timestamp(dt))
```

**Comment by MarcoGorelli** [comment]
you can do
```python
df = pd.read_csv(data, sep=",", parse_dates=['designation_date'], date_format='mixed')
```

**Comment by ciscorucinski** [comment]
Thank you for this suggestion. It is working with the first suggestion. Now just dealing with NaT values (https://github.com/pandas-dev/pandas/issues/11953), which I might be able to deal with as-is.

However, I do still think extra documentation beyond just stating the alternative method within the FutureWarning would be beneficial. I don't think the FutureWarning message needs to be modified, but there is very little guidance searching online assuming a starting point of a working date_parser implementation. But I do assume it is probably way easier than I am thinking.

**Comment by MarcoGorelli** [comment]
sure, pull requests to improve the docs would be very welcome https://pandas.pydata.org/docs/dev/development/contributing_docstring.html

**Comment by MarcoGorelli** [comment]
It's the same as the 'format' argument in 'to_datetime', a PR to clarify this would be welcome

**Comment by brendan-m-murphy** [comment]
Date parser could be used to combine multiple columns into a single datetime (e.g. columns for 'year', 'month', 'day', 'hour' minute', 'second' could be turned into a single datetime column). Is this no longer possible?

**Comment by sm-Fifteen** [comment]
> Date parser could be used to combine multiple columns into a single datetime (e.g. columns for 'year', 'month', 'day', 'hour' minute', 'second' could be turned into a single datetime column). Is this no longer possible?

@brendan-m-murphy: I believe you can still do this with `parse_dates={'full_timestamp':['my_year', 'my_month', 'my_day', 'my_hour', 'my_minute']} date_format={'full_timestamp': '%Y %m %d %H %M'}`.

**Comment by gsgxnet** [comment]
Would be great if that new parser would be able to convert POSIX times (epoch seconds). The referenced documentary for the `strftime` in Python does not say it supports `%s`.  
Linux strftime man3 manpage: 
```
       %s     The number of seconds since the Epoch, 1970-01-01 00:00:00
              +0000 (UTC). (TZ) (Calculated from mktime(tm).)
```
And sqlite seems to support it too: <https://www.sqlite.org/lang_datefunc.html>

**Comment by MarcoGorelli** [comment]
@gsgxnet could you open a new issue please?

**Comment by sm-Fifteen** [comment]
> > Date parser could be used to combine multiple columns into a single datetime (e.g. columns for 'year', 'month', 'day', 'hour' minute', 'second' could be turned into a single datetime column). Is this no longer possible?
> 
> @brendan-m-murphy: I believe you can still do this with `parse_dates={'full_datetime':['my_year', 'my_month', 'my_day', 'my_hour', 'my_minute', 'my_second']} date_format={'full_datetime': '%Y %m %d %H %M %S'}`.

I guess that's not going to hold true in the near future, since it looks like that feature is set to get deprecated in the next Pandas version, as per PR #56569 and issue #55569.
